proposals/nnnn-strict-memory-safety.md

```diff
@@ -0,0 +1,557 @@
+
+There are a few exemptions to the rule that any unsafe constructs within the signature require the declaration to be `@unsafe`:
+
+* Local variables involving unsafe types do not need to be marked with `@unsafe`. For example, the local variable `base` will have unsafe type `UnsafePointer?`, but does not require `@unsafe` because every *use* of this local variable will need to be marked using the `unsafe` expression described in the next section.
+
+  ```swift
+  func sum(array: [Int]) -> Int {
+    array.withUnsafeBufferPointer { buffer in
+      let base = /*unsafe*/ buffer.baseAddress
+      // ...
+    }
+  }
+  ```
+
+* Default arguments of functions are part of the implementation of a function, not its signature. For example, the following function does not have any unsafe types in its signature, so it does not require `@unsafe`, even though the default argument for `value` involves unsafe code. That unsafe code is effectively part of the body of the function, so it follows the rules for `unsafe` expressions.
+
+  ```swift
+  func hasDefault(value: Int = /*unsafe*/ getIntegerUnsafely()) { ... }
+  ```
+
+### `unsafe` expression
+
+When a declaration is marked `@unsafe`, it is free to use any other unsafe types as part of its interface. Any time there is executable code that makes use of unsafe constructs, that code must be within an `unsafe` expression or it will receive a diagnostic about uses of unsafe code. In the example from the previous section, `wrapper` can be marked as `@unsafe` to suppress diagnostics by explicitly propagating unsafety to their clients:
+
+```swift
+// Module B
+import A
+
+public struct MyType {
+  @unsafe public var wrapper: DataWrapper
+}
+```
+
+However, the use of the `wrapper` property in module `C` will produce a diagnostic unless it is part of an `unsafe` expression, like this:
+
+```swift
+// Module C
+import A
+import B
+
+extension MyType {
+  public func checksum() -> Int32 {}
+    return unsafe wrapper.checksum()
+  }
+}
+```
+
+The `unsafe` expression is much like `try` and `await`, in that it acknowledges that unsafe constructs (`wrapper`) are used within the subexpression but otherwise does not change the type. Unlike `try` and `await`, which require the enclosing context to handle throwing or be asynchronous, respectively, the `unsafe` expression does not imply any requirements about the enclosing block: it is purely a marker to indicate the presence of unsafe code, silencing a diagnostic.
+
+### Unsafe language constructs
+
+The following language constructs are always considered to be unsafe:
+
+* `unowned(unsafe)`: Used to store a reference without maintaining its reference count. The safe counterpart, `unowned`, uses dynamic checking to ensure that the reference isn't accessed after the corresponding object has been released. The `unsafe` variant disables that dynamic checking. Uses of `unowned(unsafe)` entities are not memory-safe.
```

Comment:
Not sure what is the policy on mentioning constructs not officially part of the language yet. 

With the upcoming lifetimes work we have quite a few including:
* `@_unsafeNonescapableResult`
* `_overrideLifetime`

.. and probably more. @atrick probably has a more exhaustive list.

Reply:
For underscored attributes, let's make the implementation Do The Right Thing. When those attributes get formalized, we should note that they are unsafe.

Reply:
`_overrideLifetime` is annotated as `@unsafe`. That's the only one I would expect to formalize.